### ISSUE TITLE
Support `warn_unused_result` on alias templates?

### ISSUE THREAD

**Issue by snarkmaster** [llvm/llvm-project#68456]
Like @dwblaikie in https://reviews.llvm.org/D102122, I am looking to annotate a very common error return type in my codebase as "[[nodiscard]]" to make it less likely that error handling gets forgotten.

However, I'm talking about a new-ish C++ codebase (actually several!), so we use templated types, analogous to this:

```
template <typename T>
using SystemMaybe = std::expected<T, std::system_error>;
```

While it's morally very similar, I can't use David's fancy new attribute here, because it's not supported syntactically -- I tried all the permutations, e.g. https://godbolt.org/z/EP34Tac7s

This leaves me with only not-great options:
 - Try to mark every function `[[nodiscard]] SystemMaybe<>` -- noisy and prone to being forgotten (clang-tidy to the rescue).
 - Define my own `expected` type with `[[nodiscard]]` composing the standard one -- very expensive in terms of the amount of boilerplate I need. Not going to do that.
 
Is it feasible to support this as a clang-specific attribute? For example, this would save the day:

```
template <typename T>
using [[clang::warn_unused_result]] SystemMaybe = std::expected<T, std::system_error>;
```

Thanks!

**Comment by shafik** [comment]
CC @AaronBallman since he was involved in the review

**Comment by dwblaikie** [comment]
Seems like a fine request - though might be the sort of thing you'd need to write a patch yourself/might not be something someone else will get to soon.

https://godbolt.org/z/o4M4faM1o

Shows that it's pretty close - the non-template `using` decl works, but the template case doesn't at the moment - seems totally reasonable/generalization that both should work just the same.

**Comment by snarkmaster** [comment]
@dwblaikie, if you think this is a change that's viable for someone completely clueless about how clang works on the inside, feel free to drop some pointers here. If you reduce it to something mechanical, I could maybe take a stab at it.

**Comment by AaronBallman** [comment]
The templated variant still has the attribute on it in the AST: https://godbolt.org/z/1avMnM9ra so my guess is that we should be looking through templated declarations [here](https://github.com/llvm/llvm-project/blob/bcf172ec578a2a787de45858871e59fd64f7b196/clang/lib/AST/Expr.cpp#L1574).

**Comment by dwblaikie** [comment]
Yep, thanks @AaronBallman - sounds like the right pointers for @snarkmaster - you can see the code there that handles TypedefDecls, and in the godbolt link from Aaron, you can see the other node types that are present for an alias template. And so some similar code that also checks for the alias template case would be good. And from my previous review/commit you can see where the tests go and add some test coverage there for the new functionality.

**Comment by philnik777** [comment]
This looks very similar to #60181.

**Comment by dwblaikie** [comment]
> This looks very similar to #60181.

Related, but at least for now they use distinct codepaths, I think. Maybe there'd be some way to unify them.
